### PR TITLE
ci: metrics: improve density tests interaction with KSM

### PR DESCRIPTION
**Pull request by grahamwhaley**
Previously we were fairly benign about our interactions with KSM - we just turned it on/off and depended on the distro default settings, whilst also hard wiring a sleep (300s) to try and ensure KSM page merging had settled down before taking our measurements.

Let's be a bit smarter by:
 - setting some aggressive KSM settings that we define
 - actively monitoring KSM to try and spot when it 'settles'

A few benefits fall out of this (on my test system):
 - the KSM density test now completes in ~123s, rather than 300s (my original goal)
 - the density result is significantly smaller - it would seem 300s was not enough to wait for KSM to really settle!
 - the result is more repeatable (CoV of ~0.3 instead of ~0.4 - it was pretty good already, and now it is even better)

### CODE REVIEW

**Comment by grahamwhaley** [comment]
@sameo - just for ref, maybe for use with some part of https://github.com/kata-containers/ksm-throttler - I used a pretty simple algo to try and spot KSM 'settling' - look at the pages merged count when the full_runs counter ticks over, and if the change is <5%, say that is good enough. It feels fairly stable and reliable in my local testing. The one downside is you have to wait for one complete full_scan cycle to spot we are settled, but that's not too bad.

**Comment by devimc** [comment]
lgtm

File: metrics/density/docker_memory_usage.sh [hunk 1]
```diff
@@ -120,6 +123,59 @@ function get_pss_memory(){
...
+
+	oldscan=$(cat /sys/kernel/mm/ksm/full_scans)
+
+	# Go around the loop until either we see a small % change
+	# between two full_scans, or we timeout
+	for ((t=0; t<$1; t++)); do
+
+		newscan=$(cat /sys/kernel/mm/ksm/full_scans)
+		if (( newscan != oldscan )); then
+			echo ""
```
Comment (MarioCarrilloA) [inline metrics/density/docker_memory_usage.sh]: this could be in order to print a newline or maybe could be a typo?

Comment (grahamwhaley) [reply metrics/density/docker_memory_usage.sh]: correct the first guess - it is to put out a newline to terminate the 'progress dots' generated by the line `echo -n "."` further down :-) I could comment it or make it more obvious with an `echo -n -e "\n"` if you want?
Or maybe better, I can change the line below to:
`echo -e "\nnew full_scan ($oldscan to $newscan)"`

let me do that one - that's probably the most succinct.

Comment (grahamwhaley) [reply metrics/density/docker_memory_usage.sh]: updated and pushed

**Comment by jodh-intel** [comment]
lgtm

**Comment by devimc** [comment]
@klynnrif could you review this pr?

File: metrics/density/README.md [hunk 2]
```diff
@@ -7,7 +7,10 @@ This number of containers could be working under different configurations/condit
 - `docker_memory_usage.sh`: measures the Proportional Set Size (PSS) memory average
    for a number of containers launched in idle mode. This test uses the `sleep` command
    to allow any memory optimizations to 'settle' (e.g. KSM execution) for a configurable
-   period of time.
+   period of time. It also has an optional 'auto' mode which will try to determine when KSM
```
Comment (klynnrif) [inline metrics/density/README.md]: Lines 10-11 suggested rewrite: This test has an optional 'auto' mode that tries to determine when KSM settles down by examining the KSM `full_scans` and `pages_shared` values.

File: metrics/density/README.md [hunk 3]
```diff
@@ -7,7 +7,10 @@ This number of containers could be working under different configurations/condit
...
+   has settled down by examining the KSM `full_scans` and `pages_shared` values. auto mode
```
Comment (klynnrif) [inline metrics/density/README.md]: Lines 11-13 suggested rewrite: Auto mode detection of KSM settling overrides the timeout value to take the measurements and to complete the test.

Comment (grahamwhaley) [reply metrics/density/README.md]: @klynnrif I've re-written that section - hopefully that is now clearer?

File: metrics/density/README.md [hunk 4]
```diff
@@ -7,7 +7,19 @@ This number of containers could be working under different configurations/condit
...
+   detection of KSM settling over-rides the timeout value for taking the measurements and
+   completing the test.
+- `docker_memory_usage.sh`: measures the Proportional Set Size (PSS) memory average
+   for a number of containers launched in idle mode. 
+   The test has two methods that can be utilized to wait for `KSM` to 'settle'.
```
Comment (klynnrif) [inline metrics/density/README.md]: Two methods exist that you can use to wait for `KSM` to 'settle'.

File: metrics/density/README.md [hunk 5]
```diff
@@ -7,7 +7,19 @@ This number of containers could be working under different configurations/condit
...
+    - The test will sleep for the supplied `timeout` argument seconds before
```
Comment (klynnrif) [inline metrics/density/README.md]: Lines 17-18 suggested rewrite: The test sleeps for the amount of time specified by 'timeout', which is in seconds, before taking measurements.

File: metrics/density/README.md [hunk 6]
```diff
@@ -7,7 +7,19 @@ This number of containers could be working under different configurations/condit
...
+      taking measurements
+    - If the `auto` mode is enabled, the test will terminate the `timeout` early if
```
Comment (klynnrif) [inline metrics/density/README.md]: Lines 19-21 suggested rewrite: If 'auto' mode is enabled, the test terminates 'timeout' early under the condition that 'KSM' has already settled, which is determined by examining the `KSM` `full_scans` and `pages_shared` information.

File: metrics/density/README.md [hunk 7]
```diff
@@ -7,7 +7,19 @@ This number of containers could be working under different configurations/condit
 - `docker_memory_usage.sh`: measures the Proportional Set Size (PSS) memory average
    for a number of containers launched in idle mode. This test uses the `sleep` command
    to allow any memory optimizations to 'settle' (e.g. KSM execution) for a configurable
-   period of time.
+   period of time. It also has an optional 'auto' mode which will try to determine when KSM
+   has settled down by examining the KSM `full_scans` and `pages_shared` values. auto mode
+   detection of KSM settling over-rides the timeout value for taking the measurements and
+   completing the test.
+- `docker_memory_usage.sh`: measures the Proportional Set Size (PSS) memory average
```
Comment (klynnrif) [inline metrics/density/README.md]: Suggest to capitalize "measures" for consistency.

File: metrics/density/README.md (hunk 3 above)
Comment (klynnrif) [reply metrics/density/README.md]: @grahamwhaley it is - I had a couple suggestions for the new section to make the voice active. Thanks!

File: metrics/density/README.md (hunk 6 above)
Comment (grahamwhaley) [reply metrics/density/README.md]: updated @klynnrif

**Pull request merged**